### PR TITLE
Introduce RpcError

### LINKED ISSUE #677: Introduce Rpc specfic errors
We should create Error subset (`class RpcError extends Error`) which should have `status: number` field.
Req/Resp should send this status field instead of determining status from error message.

### PR DESCRIPTION
**Pull request by twoeths**
## Goal
+ Close #677

### CODE REVIEW

File: packages/lodestar/src/network/encoder.ts [hunk 1]
```diff
@@ -56,7 +57,7 @@ export class ReqRespEncoder {
           }
           let status = 0;
           if(chunk.err) {
-            status = chunk.err.message === ERR_INVALID_REQ ? 1 : 2;
+            status = chunk.err.status === ERR_INVALID_REQ ? 1 : 2;
```
Comment (mpetrunic) [inline packages/lodestar/src/network/encoder.ts]: do we need this anymore?

Also add break statement so we stop yielding chunks

File: packages/lodestar/src/network/error.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/**
+ * Error of network req/resp
+ */
+
+export class RpcError extends Error {
+  public status: number;
+  constructor(message: string, status?: number) {
```
Comment (mpetrunic) [inline packages/lodestar/src/network/error.ts]: status could be enum

File: packages/lodestar/src/network/encoder.ts [hunk 3]
```diff
@@ -55,12 +55,12 @@ export class ReqRespEncoder {
               return encoder.encode(type, result);
             }, chunk.output) as Buffer;
           }
-          let status = 0;
-          if(chunk.err) {
-            status = chunk.err.status === ERR_INVALID_REQ ? 1 : 2;
-            //TODO: stop yielding
-          }
+          const status = chunk.err && chunk.err.status;
```
Comment (mpetrunic) [inline packages/lodestar/src/network/encoder.ts:58]: ain't this gonna be true/false now?

Comment (twoeths) [reply packages/lodestar/src/network/encoder.ts:58]: no, it's gonna be `undefined` or  `typeof chunk.err.status` which is number. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_Operators

File: packages/lodestar/src/constants/network.ts [hunk 4]
```diff
@@ -19,10 +19,13 @@ export enum ReqRespEncoding {
   SSZ_SNAPPY = "ssz_snappy",
 }
 
-export const ERR_INVALID_REQ = "invalid request";
-export const ERR_RESP_TIMEOUT = "response timeout";
+export enum RpcErrorEnum {
```
Comment (wemeetagain) [inline packages/lodestar/src/constants/network.ts]: Can we call this `RpcErrorType` or `RpcErrorCode`?

File: packages/lodestar/src/network/error.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import {RpcErrorEnum} from "../constants";
+
+/**
+ * Error of network req/resp
+ */
+
+export class RpcError extends Error {
+  public status: RpcErrorEnum;
+  constructor(message: string, status?: RpcErrorEnum) {
```
Comment (wemeetagain) [inline packages/lodestar/src/network/error.ts]: I think an `RpcError` will always require a status and only optionally include a message.
probably better to have the signature be `constructor(status: RpcErrorType, message?: string)`
(then we don't need the `withStatus` static method)

**Review by wemeetagain — CHANGES_REQUESTED** [review]
Looks good, only minor comments

File: packages/lodestar/src/constants/network.ts (hunk 4 above)
Comment (wemeetagain) [reply packages/lodestar/src/constants/network.ts]: bump

**Review by wemeetagain — CHANGES_REQUESTED** [review]
LGTM other than minor renaming

**Comment by twoeths** [comment]
Ready for review

**Review by wemeetagain — APPROVED** [review]
:100:

**Pull request merged**